File: symfony-app/src/Service/CommentModerationService.php
<?php

namespace App\Service;

use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
use Symfony\Contracts\HttpClient\HttpClientInterface;

class CommentModerationService
{
    /** @var array<string, array{blocked:bool, score:float, label:string, source:string}> */
    private array $cache = [];

    public function __construct(
        private readonly HttpClientInterface $httpClient,
        private readonly ForbiddenWordsFilterService $fallbackFilter
    ) {}

    /**
     * @return array{blocked:bool, score:float, label:string, source:string}
     */
    public function moderate(string $text): array
    {
        $normalized = trim($text);
        if ($normalized === '') {
            return [
                'blocked' => true,
                'score' => 1.0,
                'label' => 'EMPTY',
                'source' => 'empty',
            ];
        }

        $cacheKey = sha1(mb_strtolower($normalized));
        if (isset($this->cache[$cacheKey])) {
            return $this->cache[$cacheKey];
        }

        $remote = $this->moderateWithModel($normalized);
        if ($remote !== null) {
            return $this->cache[$cacheKey] = $remote;
        }

        $cleaned = $this->fallbackFilter->cleanText($normalized);
        $blocked = $cleaned === '' || $this->fallbackFilter->hasBadWord($cleaned);

        return $this->cache[$cacheKey] = [
            'blocked' => $blocked,
            'score' => $blocked ? 0.9 : 0.1,
            'label' => $blocked ? 'TOXIC_FALLBACK' : 'CLEAN_FALLBACK',
            'source' => 'fallback_words',
        ];
    }

    public function isInappropriate(string $text): bool
    {
        return $this->moderate($text)['blocked'];
    }

    /**
     * @return array{blocked:bool, score:float, label:string, source:string}|null
     */
    private function moderateWithModel(string $text): ?array
    {
        $endpoint = trim((string) ($_ENV['TOXICITY_MODEL_ENDPOINT'] ?? $_SERVER['TOXICITY_MODEL_ENDPOINT'] ?? ''));
        if ($endpoint === '') {
            return null;
        }

        $thresholdRaw = $_ENV['TOXICITY_THRESHOLD'] ?? $_SERVER['TOXICITY_THRESHOLD'] ?? '0.60';
        $threshold = is_numeric($thresholdRaw) ? (float) $thresholdRaw : 0.60;
        $threshold = max(0.0, min(1.0, $threshold));

        $payload = [
            'text' => $text,
            'model' => (string) ($_ENV['TOXICITY_MODEL_NAME'] ?? $_SERVER['TOXICITY_MODEL_NAME'] ?? 'unitary/toxic-bert'),
        ];

        try {
            $response = $this->httpClient->request('POST', $endpoint, [
                'json' => $payload,
                'timeout' => 4.0,
            ]);
            $data = $response->toArray(false);
        } catch (TransportExceptionInterface|\Throwable) {
            return null;
        }

        $normalized = $this->normalizeModelResponse($data);
        if ($normalized === null) {
            return null;
        }

        $label = $normalized['label'];
        $score = $normalized['score'];
        $blocked = $this->isToxicLabel($label) && $score >= $threshold;

        return [
            'blocked' => $blocked,
            'score' => $score,
            'label' => $label,
            'source' => 'toxicity_model',
        ];
    }

    /**
     * @param mixed $data
     * @return array{label:string, score:float}|null
     */
    private function normalizeModelResponse(mixed $data): ?array
    {
        if (!is_array($data)) {
            return null;
        }

        if (isset($data['result']) && is_array($data['result'])) {
            $nested = $this->normalizeModelResponse($data['result']);
            if ($nested !== null) {
                return $nested;
            }
        }

        if (array_is_list($data) && isset($data[0]) && is_array($data[0])) {
            $nested = $this->normalizeModelResponse($data[0]);
            if ($nested !== null) {
                return $nested;
            }
        }

        $label = isset($data['label']) && is_string($data['label'])
            ? strtoupper(trim($data['label']))
            : null;

        $score = null;
        if (isset($data['score']) && is_numeric($data['score'])) {
            $score = (float) $data['score'];
        } elseif (isset($data['confidence']) && is_numeric($data['confidence'])) {
            $score = (float) $data['confidence'];
        }

        if ($label === null || $score === null) {
            return null;
        }

        $score = $score > 1.0 ? ($score / 100.0) : $score;
        $score = max(0.0, min(1.0, $score));

        return ['label' => $label, 'score' => $score];
    }

    private function isToxicLabel(string $label): bool
    {
        if (str_contains($label, 'NON_TOXIC') || str_contains($label, 'NOT_TOXIC') || str_contains($label, 'CLEAN')) {
            return false;
        }

        $toxicityHints = ['TOXIC', 'INSULT', 'OBSCENE', 'PROFAN', 'HATE', 'ABUSE', 'OFFENSIVE', 'LABEL_1'];
        foreach ($toxicityHints as $hint) {
            if (str_contains($label, $hint)) {
                return true;
            }
        }

        return false;
    }

}

File: javafx-app/src/main/java/com/santea/service/CommentModerationService.java
package com.santea.service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.util.HexFormat;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Modération automatique : modèle toxicité distant + fallback mots interdits
 * (symfony-app CommentModerationService).
 */
public class CommentModerationService {
    private static final Duration HTTP_TIMEOUT = Duration.ofSeconds(4);

    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(HTTP_TIMEOUT)
            .build();

    private final ForbiddenWordsFilterService fallbackFilter;
    private final Map<String, ModerationOutcome> cache = new ConcurrentHashMap<>();

    public CommentModerationService(ForbiddenWordsFilterService fallbackFilter) {
        this.fallbackFilter = fallbackFilter;
    }

    public ModerationOutcome moderate(String text) {
        String normalized = text == null ? "" : text.trim();
        if (normalized.isEmpty()) {
            return new ModerationOutcome(true, 1.0, "EMPTY", "empty");
        }

        String cacheKey = sha1Hex(normalized.toLowerCase(Locale.ROOT));
        ModerationOutcome cached = cache.get(cacheKey);
        if (cached != null) {
            return cached;
        }

        ModerationOutcome remote = moderateWithModel(normalized);
        if (remote != null) {
            cache.put(cacheKey, remote);
            return remote;
        }

        String cleaned = fallbackFilter.cleanText(normalized);
        boolean blocked = cleaned.isEmpty() || fallbackFilter.hasBadWord(cleaned);
        ModerationOutcome out = new ModerationOutcome(
                blocked,
                blocked ? 0.9 : 0.1,
                blocked ? "TOXIC_FALLBACK" : "CLEAN_FALLBACK",
                "fallback_words"
        );
        cache.put(cacheKey, out);
        return out;
    }

    public boolean isInappropriate(String text) {
        return moderate(text).blocked();
    }

    private ModerationOutcome moderateWithModel(String text) {
        String endpoint = getenv("TOXICITY_MODEL_ENDPOINT");
        if (endpoint.isEmpty()) {
            return null;
        }

        double threshold = parseThreshold(getenv("TOXICITY_THRESHOLD"), 0.60);
        String modelName = getenv("TOXICITY_MODEL_NAME");
        if (modelName.isEmpty()) {
            modelName = "unitary/toxic-bert";
        }

        String body = CommunityHttpJson.buildTextModelBody(text, modelName);
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(endpoint))
                    .timeout(HTTP_TIMEOUT)
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                    .build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                return null;
            }
            LabelScore parsed = normalizeModelResponse(response.body());
            if (parsed == null) {
                return null;
            }
            boolean blocked = isToxicLabel(parsed.label()) && parsed.score() >= threshold;
            return new ModerationOutcome(blocked, parsed.score(), parsed.label(), "toxicity_model");
        } catch (Exception e) {
            return null;
        }
    }

    private LabelScore normalizeModelResponse(String raw) {
        if (raw == null || raw.isBlank()) {
            return null;
        }
        String slice = CommunityHttpJson.unwrapNestedResponse(raw);
        String labelStr = CommunityHttpJson.findStringField(slice, "label");
        Double score = CommunityHttpJson.findNumericField(slice, "score");
        if (score == null) {
            score = CommunityHttpJson.findNumericField(slice, "confidence");
        }
        if (labelStr == null || score == null) {
            return null;
        }
        String label = labelStr.toUpperCase(Locale.ROOT).trim();
        double s = score;
        if (s > 1.0) {
            s = s / 100.0;
        }
        s = Math.max(0.0, Math.min(1.0, s));
        return new LabelScore(label, s);
    }

    private boolean isToxicLabel(String label) {
        String u = label.toUpperCase(Locale.ROOT);
        if (u.contains("NON_TOXIC") || u.contains("NOT_TOXIC") || u.contains("CLEAN")) {
            return false;
        }
        String[] hints = {"TOXIC", "INSULT", "OBSCENE", "PROFAN", "HATE", "ABUSE", "OFFENSIVE", "LABEL_1"};
        for (String hint : hints) {
            if (u.contains(hint)) {
                return true;
            }
        }
        return false;
    }

    private static double parseThreshold(String raw, double defaultValue) {
        if (raw == null || raw.isBlank()) {
            return defaultValue;
        }
        try {
            double t = Double.parseDouble(raw.trim());
            return Math.max(0.0, Math.min(1.0, t));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private static String getenv(String key) {
        String v = System.getenv(key);
        return v == null ? "" : v.trim();
    }

    private static String sha1Hex(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            byte[] digest = md.digest(input.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(digest);
        } catch (Exception e) {
            return Integer.toHexString(input.hashCode());
        }
    }

    private record LabelScore(String label, double score) {
    }

    public record ModerationOutcome(boolean blocked, double score, String label, String source) {
    }
}
